### PR TITLE
Proposes the use of Python's Dataclasses

### PR DESCRIPTION
**Pull request by verdan**
Signed-off-by: verdan <verdan.mahmood@gmail.com>

Python 3.7 introduced [dataclasses (PEP557)](https://docs.python.org/3/library/dataclasses.html). Dataclasses can be a convenient way to generate classes whose primary goal is to contain values. 

We are relying on `attrs` for (de-)serializing data for each request within Amundsen. `attrs` itself is a big package with so many different features available that we don't even need. 

By using Python's native dataclasses we can have one less dependency with almost the same features `attrs` provide that we intend to use.

### CODE REVIEW

**Comment by feng-tao** [comment]
currently Amundsen supports python36 and above. Will it force to drop python36 support?

**Comment by verdan** [comment]
> currently Amundsen supports python36 and above. Will it force to drop python36 support?

@feng-tao Yes, it will. I've made it explicit in the technical details and also in the drawbacks. 

I believe we should aim to support as many python versions as possible, but MUST support at least the last 3 versions. The current Python version is 3.9, so I think it is okay to drop the support of Python3.6.

**Comment by feng-tao** [comment]
Given python36 is still widely used these days while python39 is just introduced, I will be hesitant to just drop python36. I take a quick look at airflow and superset which both still supports py3.6.

**Comment by feng-tao** [comment]
@verdan actually it is backported to py36 (https://github.com/ericvsmith/dataclasses). I think it is fine from the python version pov.

File: rfcs/000-python-dataclasses.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+complex, instead look for a simpler and cleaner solution. 
+
+
+## Prior art
+N/A
+
+## Unresolved questions
+
+- Should we aim for having as less dependencies as possible? That includes the third party packages. 
+- Do we think of a case where we'll be doing heavy data validations, and if those can not be done with dataclasses?
```
Comment (feng-tao) [inline rfcs/000-python-dataclasses.md:84]: I think it is ok.

**Comment by verdan** [comment]
> @verdan actually it is backported to py36 (https://github.com/ericvsmith/dataclasses). I think it is fine from the python version pov.

oh, this is nice. I did not know that before. I am still curious however why companies are still using Python3.6, as Python3.7 introduced a number of nice feature that the community was waiting for. but yes, the major share is still Python3.6 as you mentioned.

**Comment by dorianj** [comment]
Strongly in support of this -- cleaner syntax and typing, and the python3.6 backport works well. Any reason not to push this through and start using?

**Comment by verdan** [comment]
@feng-tao do we agree to move this to the `Active` state, and start implementing this if no further comments?

**Comment by feng-tao** [comment]
sgtm

**Comment by jornh** [comment]
To quote https://github.com/amundsen-io/rfcs/pull/12#issuecomment-725277808

> I believe we should aim to support as many python versions as possible, but MUST support at least the last 3 versions. The current Python version is 3.9.

Not really related to this RFC itself, but since the topic is up. We're currently only officially supporting 3.6 and 3.7. Looks like the Databuilder package works for folks with 3.8, but tests need some love + maybe addition to CI. Similarly the 3 services need some work just to be able to build with 3.8.

Time to kick off that work? I'd be happy to do some of it. I guess it doesn't need a RFC as it's basically just a maintenance chore, not introducing any new feature or breaking anything.

File: rfcs/000-python-dataclasses.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+## UI/UX-level Explanation
+
+Nothing changes on the UI/UX.
+
+## Reference-level Explanation (aka Technical Details)
+
+- The primary change will happen in amundsencommon repositories where we are defining the models. I believe no other major change 
+will be required in any other repository. 
+
+- DROPS SUPPORT OF PYTHON 3.6
```
Comment (jornh) [inline rfcs/000-python-dataclasses.md:60]:
```suggestion
- Python 3.6 can still be supported as long as it's needed through  backported to py36 ([ericvsmith/dataclasses](https://github.com/ericvsmith/dataclasses)). 
```

File: rfcs/000-python-dataclasses.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+## Drawbacks
+
+- One of the main thing to consider is the Python 3.6 support. If we go for the dataclasses, we will not be supporting 
+Python 3.6 anymore. The minimum version requirement for Python for Amundsen will become 3.7.
+
```
Comment (jornh) [inline rfcs/000-python-dataclasses.md:66]:
```suggestion
```

**Comment by verdan** [comment]
> To quote [#12 (comment)](https://github.com/amundsen-io/rfcs/pull/12#issuecomment-725277808)
> 
> > I believe we should aim to support as many python versions as possible, but MUST support at least the last 3 versions. The current Python version is 3.9.
> 
> Not really related to this RFC itself, but since the topic is up. We're currently only officially supporting 3.6 and 3.7. Looks like the Databuilder package works for folks with 3.8, but tests need some love + maybe addition to CI. Similarly the 3 services need some work just to be able to build with 3.8.
> 
> Time to kick off that work? I'd be happy to do some of it. I guess it doesn't need a RFC as it's basically just a maintenance chore, not introducing any new feature or breaking anything.

@jornh that would be awesome if you could please help out with making sure that all the services work with the latest 3 versions of Python. 

In the meantime, I am taking care of the dataclasses.

**Comment by dorianj** [comment]
Note that dataclasses were landed in the front-end repo, with the back-compat library for 3.6, and is working smoothly.

**Review by dorianj — APPROVED** [review]

**Pull request merged**